app/src/main/java/com/example/firstapplication/MainActivity.kt

```diff
@@ -7,28 +7,30 @@ import java.util.*
 
 class MainActivity : AppCompatActivity() {
     private lateinit var binding: ActivityMainBinding
+    private lateinit var usersViewModel: UserListViewModel
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        val userAdapter = UserAdapter(
-            mutableListOf(
-                User("user-id-1", "user-name-1", "user-1-email"),
-                User("user-id-2", "user-name-2", "user-2-email"),
-                User("user-id-3", "user-name-3", "user-3-email"),
-                User("user-id-4", "user-name-4", "user-4-email"),
-                User("user-id-5", "user-name-5", "user-5-email")
-            )
-        )
         val recyclerView = binding.recyclerView
-        recyclerView.adapter = userAdapter
 
+        usersViewModel = UserListViewModel()
+        usersViewModel.users.observe(
+            this,
+            { updatedUserList ->
+                recyclerView.adapter = UserAdapter(updatedUserList.toMutableList())
```

Comment:
Nie nie :) Nie tworzymy nowego adapteru za każdym razem gdy zmodyfikujemy listę :) tylko wstawmy listę do raz stworzonego adapteru :) pewnie trzeba będzie wywołać w adapterze notifyItemsChanged(). Jest to mniej wydajne od bardziej wyspecjalizowanych metod jak notifyItemInserted() czy notifyItemRemoved() ale jeśli chcesz przepychać całą listę po każdej zmianie, to jest to konieczne. Kilka lekcji dalej poznasz lepszy mechanizm  (DiffUtil).

Reply:
Miałem na myśli **notifyDataSetChanged**